### PR TITLE
DAOS-8870 pool: Allow srv_hdls in TGT_QUERY_MAP

### PR DESCRIPTION
**Pull request by liw**
ds_pool_tgt_query_map_handler misses (due to my ignorance) the case
where an engine may go through the map_refresh code to refresh a
"client-copy" of the pool map during a rebuild job. This patch fixes the
issue by following the handling "server handles" in
ds_pool_query_handler. Also,

  - Extract a new ds_pool_hdl_is_from_srv API for
    ds_pool_tgt_query_map_handler, to avoid looking up the ds_pool
    object twice.
  - Handle -DER_NOTLEADER from ds_pool_hdl_is_from_srv in
    ds_pool_tgt_query_map_handler and resulting -DER_AGAIN in
    map_refresh_cb.
  - Remove the is_pool_from_srv declaration from daos_engine.h.

Signed-off-by: Li Wei <wei.g.li@intel.com>

### CODE REVIEW

**Comment by liw** [comment]
> Test stage Functional Hardware Large completed with status FAILURE. https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-7082/3/display/redirect

Due to a post-provision failure.

File: src/pool/srv_pool.c [hunk 1]
```diff
@@ -5701,11 +5703,33 @@ int ds_pool_prop_fetch(struct ds_pool *pool, unsigned int bits,
...
+ds_pool_hdl_is_from_srv(struct ds_pool *pool, uuid_t hdl)
+{
+	uuid_t	srv_hdl;
+	int	rc;
+
+	rc = ds_pool_iv_srv_hdl_fetch(pool, &srv_hdl, NULL);
+	if (rc != 0)
+		return rc;
+
+	return uuid_compare(pool->sp_uuid, hdl) == 0;
```
Comment (liw) [inline src/pool/srv_pool.c:5726]: Oops, "pool->sp_uuid" is a typo; should be "srv_hdl".

**Review by liuxuezhao — APPROVED** [review]

**Comment by liw** [comment]
@wangdi1, I was asked again about this ticket. Would you mind replying if this is no longer required or if you're able to review this, please?

File: src/pool/srv_target.c [hunk 2]
```diff
@@ -1377,42 +1377,72 @@ ds_pool_tgt_query_map_handler(crt_rpc_t *rpc)
...
+		 * See the comment on validating the pool handle in
+		 * ds_pool_query_handler.
+		 */
+		pool = ds_pool_lookup(in->tmi_op.pi_uuid);
+		if (pool == NULL) {
+			D_ERROR(DF_UUID": failed to look up pool\n", DP_UUID(in->tmi_op.pi_uuid));
+			rc = -DER_NONEXIST;
+			goto out;
+		}
+		rc = ds_pool_hdl_is_from_srv(pool, in->tmi_op.pi_hdl);
```
Comment (wangdi1) [inline src/pool/srv_target.c:1411]: you can try to use sp_srv_pool_hdl in ds_pool first? instead of issuing IV here?

File: src/pool/srv_target.c [hunk 3]
```diff
@@ -1377,42 +1377,72 @@ ds_pool_tgt_query_map_handler(crt_rpc_t *rpc)
...
+		hdl = ds_pool_hdl_lookup(in->tmi_op.pi_hdl);
+		if (hdl == NULL) {
+			rc = -DER_NO_HDL;
+			goto out;
+		}
+		ds_pool_get(hdl->sph_pool);
+		pool = hdl->sph_pool;
+		ds_pool_hdl_put(hdl);
+	} else {
+		/*
```
Comment (wangdi1) [inline src/pool/srv_target.c:1401]: As we discussed on that ticket, this should not happen even for rebuild case. But probably needed it in the future, so no harm to add it. 

Though probably worth a D_INFO here.

**Review by wangdi1 — APPROVED** [review]

File: src/pool/srv_target.c (hunk 2 above)
Comment (liw) [reply src/pool/srv_target.c:1411]: @wangdi1, do you mean something like the following? I struggle to understand whether pool->sp_srv_pool_hdl can be stale (i.e., generated by a previous PS leader and different from the one generated by the current PS leader).

    --- a/src/pool/srv_pool.c
    +++ b/src/pool/srv_pool.c
    @@ -5716,6 +5716,9 @@ ds_pool_hdl_is_from_srv(struct ds_pool *pool, uuid_t hdl)
            uuid_t  srv_hdl;
            int     rc;
     
    +       if (!uuid_is_null(pool->sp_srv_pool_hdl))
    +               return uuid_compare(pool->sp_srv_pool_hdl, hdl) == 0;
    +
            rc = ds_pool_iv_srv_hdl_fetch(pool, &srv_hdl, NULL);
            if (rc != 0)
                    return rc;

Comment (liw) [reply src/pool/srv_target.c:1411]: @wangdi1, any thought on why sp_srv_pool_hdl can never be stale?

File: src/pool/srv_target.c (hunk 3 above)
Comment (liw) [reply src/pool/srv_target.c:1403]: That might be noisy; I added an error message below instead, to make sure every abnormal branching in the "server pool handle" case will print a message.

File: src/pool/srv_target.c (hunk 2 above)
Comment (wangdi1) [reply src/pool/srv_target.c:1413]: @liw sp_srv_pool_hdl is generated by PS leader during step up process, then broadcast to all servers  by IV.  For stable, you mean consistent cross the PS leader change? This seems unlikely, unless we put it to RDB.

Comment (wangdi1) [reply src/pool/srv_target.c:1413]: oh, I mean

if (uuid_compare(pool->sp_srv_pool_hdl, hdl) != 0) {
          rc = ds_pool_hdl_is_from_srv(pool, pi_hdl);

          /* extra check */
}

Or even move this uuid_compare(pool->sp_srv_pool_hdl, hdl) inside ds_pool_hdl_is_from_srv(),  so it does not have to create extra ULT each time?

**Review by wangdi1 — APPROVED** [review]

**Comment by wangshilong** [comment]
It will be nice to remove skipForTicket as well, to make sure pr could fix issue reported:

[root@server erasurecode]# git grep DAOS-8870
rebuild_fio.py:    @skipForTicket("DAOS-8870")

**Comment by liw** [comment]
> It will be nice to remove skipForTicket as well, to make sure pr could fix issue reported:
> 
> [root@server erasurecode]# git grep DAOS-8870 rebuild_fio.py: @skipForTicket("DAOS-8870")

Indeed. Thanks. I've removed this skipForTicket.

**Comment by daltonbohning** [comment]
> > It will be nice to remove skipForTicket as well, to make sure pr could fix issue reported:
> > [root@server erasurecode]# git grep DAOS-8870 rebuild_fio.py: @skipForTicket("DAOS-8870")
> 
> Indeed. Thanks. I've removed this skipForTicket.

The test is only ran for `full_regression`
```
:avocado: tags=all,full_regression
:avocado: tags=hw,large,ib2
:avocado: tags=ec,ec_array,fio,ec_online_rebuild
:avocado: tags=ec_online_rebuild_fio
```
So you would need to include the `Test-tag` commit pragma to run this test
```
Test-tag: pr ec_online_rebuild_fio
```

**Comment by liw** [comment]
> Test-tag: pr ec_online_rebuild_fio

Thank you. This is very helpful. Added.

**Comment by daltonbohning** [comment]
@liw FYI, there is also https://daosio.atlassian.net/browse/DAOS-9785. There might be a test issue with detecting when aggregation starts. Would you expect all variants of `ec_online_rebuild_fio` to pass with this PR?

**Comment by liw** [comment]
> @liw FYI, there is also https://daosio.atlassian.net/browse/DAOS-9785. There might be a test issue with detecting when aggregation starts. Would you expect all variants of `ec_online_rebuild_fio` to pass with this PR?

@daltonbohning, this patch is for future use; DAOS-8870's failures are supposed to have already been addressed by prior patches on that ticket.

**Review by wangdi1 — DISMISSED** [review]

**Comment by liw** [comment]
"Node post-provision" failures...

**Comment by liw** [comment]
Dropped

    Test-tag: pr ec_online_rebuild_fio

**Review by liuxuezhao — APPROVED** [review]

**Review by wangdi1 — APPROVED** [review]

**Pull request merged**